Nama    : Geugeut Nyarikawanti Surahmat
NIM     : 103132400002
Kelas   : S1-SD-04-01

--------------------------------------- Preprocessing Data ----------------------------------------

In [ ]:
#import libraries
import pandas as pd
import numpy as np

In [ ]:
features = pd.read_csv('features.csv')
labels = pd.read_csv('labels.csv')

1. Tampilkan data yang telah diunduh, kemudian jelaskan semua kolom yang terdapat pada dataset tersebut.

In [ ]:
#Menampilkan data "features" yang telah diunduh
features

Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  9500.0                          day   
3                   1.0     NaN                          day   
4                   1.0     NaN                          day   
...                 ...     ...                          ...   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                        1.0                             NaN  Portuguese  ...   
1                        1.0                           136.0  Portuguese  ...   
2                        1.0                           120.0  Portuguese  ...   
3                        1.0                           141.0  Portuguese  ...   
4                       39.0                           120.0     Spanish  ...   
...                      ...                             ...         ...  ...   
3091                     1.0                           140.0  Portuguese  ...   
3092                     1.0                           126.0  Portuguese  ...   
3093                     1.0                           141.0  Portuguese  ...   
3094                     4.0                           150.0  Portuguese  ...   
3095                     1.0                           118.0  Portuguese  ...   

      Curricular units 1st sem (without evaluations)  \
0                                                0.0   
1                                                0.0   
2                                                0.0   
3                                                0.0   
4                                                0.0   
...                                              ...   
3091                                             0.0   
3092                                             0.0   
3093                                             0.0   
3094                                             0.0   
3095                                             NaN   

      Curricular units 2nd sem (credited)  \
0                                     0.0   
1                                     0.0   
2                                     4.0   
3                                     0.0   
4                                     1.0   
...                                   ...   
3091                                  1.0   
3092                                  0.0   
3093                                  0.0   
3094                                  2.0   
3095                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   

In [ ]:
#Menampilkan informasi mengenai setiap kolom pada "features" 
features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3096 entries, 0 to 3095
Data columns (total 38 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Unnamed: 0                                      3096 non-null   int64  
 1   Student Id                                      3096 non-null   int64  
 2   Marital status                                  3012 non-null   object 
 3   Application mode                                2924 non-null   float64
 4   Application order                               2988 non-null   float64
 5   Course                                          2944 non-null   float64
 6   Daytime/evening attendance	                     3016 non-null   object 
 7   Previous qualification                          2973 non-null   float64
 8   Previous qualification (grade)                  3011 non-null   float64
 9   Nacionality                              

Penjelasan mengenai data features:
1. Data features memiliki 3096 baris dan 38 kolom
2. Terdapat 3 jenis tipe data, yaitu 33 kolom bertipe float64, 2 kolom bertipe int64, 3 kolom bertipe object

In [ ]:
#Menampilkan data "labels" yang telah diunduh
labels

Target
0     Graduate
1     Graduate
2     Graduate
3     Graduate
4      Dropout
...        ...
3091  Graduate
3092  Graduate
3093  Graduate
3094   Dropout
3095  Graduate

[3096 rows x 1 columns]

In [ ]:
#Menampilkan informasi mengenai setiap kolom pada "features" 
labels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3096 entries, 0 to 3095
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Target  3096 non-null   object
dtypes: object(1)
memory usage: 24.3+ KB


Penjelasan mengenai data labels:
1. Data labels memiliki 3096 baris dan 1 kolom bernama Target
2. Kolom Target bertipe data object

2. Lakukan agregasi antara tabel features dengan tabel labels.

In [ ]:
#melakukan penggabungan data "features" dengan "labels" dengan isi labels menjadi kolom ke 39
df = pd.concat([features, labels], axis=1)
df

Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  9500.0                          day   
3                   1.0     NaN                          day   
4                   1.0     NaN                          day   
...                 ...     ...                          ...   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                        1.0                             NaN  Portuguese  ...   
1                        1.0                           136.0  Portuguese  ...   
2                        1.0                           120.0  Portuguese  ...   
3                        1.0                           141.0  Portuguese  ...   
4                       39.0                           120.0     Spanish  ...   
...                      ...                             ...         ...  ...   
3091                     1.0                           140.0  Portuguese  ...   
3092                     1.0                           126.0  Portuguese  ...   
3093                     1.0                           141.0  Portuguese  ...   
3094                     4.0                           150.0  Portuguese  ...   
3095                     1.0                           118.0  Portuguese  ...   

      Curricular units 2nd sem (credited)  \
0                                     0.0   
1                                     0.0   
2                                     4.0   
3                                     0.0   
4                                     1.0   
...                                   ...   
3091                                  1.0   
3092                                  0.0   
3093                                  0.0   
3094                                  2.0   
3095                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   
...                                   ...   
3091                                  6.0   
3092                                  6.0   
3093                                  8.0   
3094                                  8.0   
3095                                  6.0   

      Curricular units 2nd sem (evaluations)  \
0                                        6.0   
1                                        0.0   
2                                        8.0   
3                                        7.0   
4                                        NaN   
...                                      ...   
3091                                     9.0   
3092             

Penjelasan:
Agregasi merupakan sebuah proses menggabungkan dua dataset. Pada praktik kali ini data labels digabungkan dengan data features sebagai kolom ke 39. Dataset baru hasil agregasi features dan labels diberi nama df

3. Tampilkan tipe data dari setiap kolom yang terdapat dalam dataset.

In [ ]:
df.dtypes

Unnamed: 0                                          int64
Student Id                                          int64
Marital status                                     object
Application mode                                  float64
Application order                                 float64
Course                                            float64
Daytime/evening attendance\t                       object
Previous qualification                            float64
Previous qualification (grade)                    float64
Nacionality                                        object
Mother's qualification                            float64
Father's qualification                            float64
Mother's occupation                               float64
Father's occupation                               float64
Admission grade                                   float64
Displaced                                         float64
Educational special needs                         float64
Debtor        

4. Tampilkan statistik deskriptif dari dataset dan berikan analisis singkat mengenai hasilnya.

In [ ]:
df.describe().T

count         mean  \
Unnamed: 0                                      3096.0  1547.500000   
Student Id                                      3096.0  5470.623062   
Application mode                                2924.0    18.785568   
Application order                               2988.0     1.734940   
Course                                          2944.0  8817.929348   
Previous qualification                          2973.0     4.685503   
Previous qualification (grade)                  3011.0   132.647293   
Mother's qualification                          2951.0    19.470010   
Father's qualification                          3000.0    22.408667   
Mother's occupation                             3034.0    11.495715   
Father's occupation                             2952.0    11.597900   
Admission grade                                 2955.0   126.902876   
Displaced                                       2974.0     0.546066   
Educational special needs                       2977.0     0.010749   
Debtor                                          2980.0     0.117450   
Tuition fees up to date                         2940.0     0.880612   
Gender                                          2894.0     0.350726   
Scholarship holder                              2930.0     0.246416   
Age at enrollment                               2892.0    23.310166   
International                                   2995.0     0.025042   
Curricular units 1st sem (credited)             2888.0     0.698407   
Curricular units 1st sem (enrolled)             2953.0     6.251270   
Curricular units 1st sem (evaluations)          3029.0     8.233080   
Curricular units 1st sem (approved)             2912.0     4.732143   
Curricular units 1st sem (grade)                3007.0    10.692416   
Curricular units 1st sem (without evaluations)  3023.0     0.133311   
Curricular units 2nd sem (credited)             3035.0     0.516639   
Curricular units 2nd sem (enrolled)             2997.0     6.208876   
Curricular units 2nd sem (evaluations)          2889.0     8.061613   
Curricular units 2nd sem (approved)             2945.0     4.422411   
Curricular units 2nd sem (grade)                3021.0    10.275115   
Curricular units 2nd sem (without evaluations)  2943.0     0.139314   
Unemployment rate                               2960.0    11.549561   
Inflation rate                                  2929.0     1.215534   
GDP                                             3032.0     0.028278   

                                                        std      min      25%  \
Unnamed: 0                                       893.882543     0.00   773.75   
Student Id                                      2582.483019  1000.00  3288.50   
Application mode                                  17.532674     1.00     1.00   
Application order                                  1.327527     1.00     1.00   
Course                                          2148.407432    33.00  9085.00   
Previous qualification                            10.439709     1.00     1.00   
Previous qualification (grade)                    13.322662    95.00   124.00   
Mother's qualification                            15.636952     1.00     2.00   
Father's qualification                            15.291008     1.00     3.00   
Mother's occupation                               28.173386     0.00     4.00   
Father's occupation                               26.799023     0.00     4.00   
Admission grade                                   14.460283    95.00   118.00   
Displaced                                          0.497957     0.00     0.00   
Educational special needs                          0.103136     0.00     0.00   
Debtor                                             0.322009     0.00     0.00   
Tuition fees up to date                            0.324299     0.00     1.00   
Gender                                             0.477280     0.00     0.00   
Scholarship holder                

Analisis singkat:
1. Terdapat missing value pada data tersebut karena ada kolom dengan count kurang dari 3096 sehingga data harus melewati proses imputasi
2. Terdapat beberapa fitur yang perlu distandarkan, seperti Course dan Grade
3. Beberapa kolom berpotensi tinggi memiliki outlier, terutama Occupation

5. Slicing Data

In [ ]:
# 20 baris awal dari dataset
df.head(20)

Unnamed: 0  Student Id Marital status  Application mode  \
0            0        6634         single               1.0   
1            1        6459         single               1.0   
2            2        2238         single              42.0   
3            3        2479         single               1.0   
4            4        4260         single              44.0   
5            5        3544        married              39.0   
6            6        4956         single              39.0   
7            7        6758         single               1.0   
8            8        2484         single               1.0   
9            9        9845         single              39.0   
10          10        3690         single               1.0   
11          11        2585         single               1.0   
12          12        5241         single               1.0   
13          13        2972         single               1.0   
14          14        2003        married              39.0   
15          15        3007         single               7.0   
16          16        9857         single              17.0   
17          17        3327         single               NaN   
18          18        5235         single              17.0   
19          19        3969         single               1.0   

    Application order  Course Daytime/evening attendance\t  \
0                 3.0  9773.0                          day   
1                 1.0     NaN                          day   
2                 1.0  9500.0                          day   
3                 1.0     NaN                          day   
4                 1.0     NaN                          day   
5                 1.0  9853.0                          day   
6                 1.0  9085.0                          day   
7                 4.0  9670.0                          day   
8                 1.0  9773.0                          day   
9                 1.0  9670.0                          day   
10                1.0  9670.0                          day   
11                1.0  9070.0                          day   
12                1.0   171.0                          day   
13                3.0  9500.0                          day   
14                1.0  8014.0                          NaN   
15                1.0  9119.0                          day   
16                1.0  9119.0                          day   
17                1.0  9003.0                          day   
18                1.0   171.0                          day   
19                1.0  9085.0                          day   

    Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                      1.0                             NaN  Portuguese  ...   
1                      1.0                           136.0  Portuguese  ...   
2                      1.0                           120.0  Portuguese  ...   
3                      1.0                           141.0  Portuguese  ...   
4                     39.0                           120.0     Spanish  ...   
5                      1.0                             NaN  Portuguese  ...   
6                      1.0                           133.1         NaN  ...   
7                      1.0                           130.0  Portuguese  ...   
8                      1.0                           135.0  Portuguese  ...   
9                      1.0                           120.0  Portuguese  ...   
10                     1.0                           133.1  Portuguese  ...   
11                     1.0                           117.0  Portuguese  ...   
12                     1.0                           162.0  Portuguese  ...   
13                     1.0                           135.0  Portuguese  ...   
14                     NaN                           120.0  Portuguese  ...   
15                     1.0                           140.8  Portuguese  ...   
16                     1.0        

In [ ]:
#20 baris terakhir dari dataset
df.tail(20)

Unnamed: 0  Student Id Marital status  Application mode  \
3076        3076        6757         single              17.0   
3077        3077        5750         single               1.0   
3078        3078        2599         single               1.0   
3079        3079        3249         single              17.0   
3080        3080        3366       divorced              39.0   
3081        3081        3462         single              39.0   
3082        3082        9194         single              17.0   
3083        3083        5574         single               1.0   
3084        3084        8311         single               1.0   
3085        3085        2304         single              17.0   
3086        3086        6615         single              43.0   
3087        3087        3759         single               1.0   
3088        3088        8085         single              51.0   
3089        3089        1573         single              17.0   
3090        3090        9737         single              51.0   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
3076                NaN  9254.0                          day   
3077                2.0  9500.0                          day   
3078                1.0  9085.0                          day   
3079                1.0  9238.0                          day   
3080                1.0  9003.0                          day   
3081                1.0   171.0                          day   
3082                5.0  9147.0                          day   
3083                1.0  9500.0                          day   
3084                2.0  9773.0                          day   
3085                1.0  9238.0                          day   
3086                1.0  9556.0                          day   
3087                1.0   171.0                          day   
3088                1.0  9670.0                          day   
3089                2.0   171.0                          day   
3090                1.0  9070.0                          day   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
3076                     1.0                           122.0  Portuguese  ...   
3077                     1.0                           140.0  Portuguese  ...   
3078                     1.0                           131.0  Portuguese  ...   
3079                     1.0                           120.0  Portuguese  ...   
3080                    12.0                           133.1  Portuguese  ...   
3081                     1.0                           147.0  Portuguese  ...   
3082                     1.0                           131.0  Portuguese  ...   
3083                     1.0                           140.0  Portuguese  ...   
3084                     1.0                           130.0  Portuguese  ...   
3085                     NaN                           139.0  Portuguese  ...   
3086                     1.0                           149.0  Portuguese  ...   
3087                     3.0                           116.0  Portuguese  ...   
3088                     1.0                           133.1  Portuguese  ...   
3089                     1.0                           142.0  Portuguese  ...   
3090                     1.0                           143.0  Portuguese  ...   

In [ ]:
# Baris dari indexnya 80-100
df.loc[80:100]

Unnamed: 0  Student Id Marital status  Application mode  \
80           80        1238         single              17.0   
81           81        4121         single              39.0   
82           82        2890         single               1.0   
83           83        4358         single               1.0   
84           84        4896         single               NaN   
85           85        4735         single              18.0   
86           86        1047         single               1.0   
87           87        7430         single              39.0   
88           88        8527         single               1.0   
89           89        5420         single              43.0   
90           90        5625         single              44.0   
91           91        8990            NaN               1.0   
92           92        9143        married              44.0   
93           93        1944        married               7.0   
94           94        3666         single               1.0   
95           95        6960         single              17.0   
96           96        7620         single               1.0   
97           97        6083         single              42.0   
98           98        5525         single              17.0   
99           99        6029         single              17.0   
100         100        9749         single              39.0   

     Application order  Course Daytime/evening attendance\t  \
80                 1.0  9670.0                          day   
81                 1.0  9003.0                          day   
82                 1.0   171.0                          day   
83                 1.0  9147.0                          day   
84                 1.0  9991.0                      evening   
85                 1.0  9500.0                          day   
86                 1.0  9773.0                          day   
87                 2.0  9556.0                          day   
88                 2.0  9500.0                          day   
89                 1.0     NaN                          day   
90                 1.0  9003.0                          day   
91                 1.0  9500.0                          day   
92                 1.0  9085.0                          day   
93                 1.0  9070.0                          day   
94                 3.0  9070.0                          day   
95                 3.0  9119.0                          day   
96                 1.0  9500.0                          day   
97                 1.0  9500.0                          day   
98                 5.0  9238.0                          day   
99                 1.0  9773.0                          day   
100                1.0  9085.0                          day   

     Previous qualification  Previous qualification (grade) Nacionality  ...  \
80                      1.0                           155.0  Portuguese  ...   
81                      1.0                             NaN  Portuguese  ...   
82                      1.0                           159.0         NaN  ...   
83                      1.0                           150.0   Ukrainian  ...   
84                      1.0                           140.0  Portuguese  ...   
85                      1.0                           144.0  Portuguese  ...   
86                      1.0                           117.0  Portuguese  ...   
87                     19.0                           133.1  Portuguese  ...   
88                      1.0                           145.0  Portuguese  ...   
89                      1.0                           121.0  Portuguese  ...   
90                     39.0                           140.0  Portuguese  ...   
91                      1.0                           141.0  Portuguese  ...   
92                     39.0                           160.0  Portuguese  ...   
93                     43.0                           140.0  Portuguese  ...   
94       

6. Tangani nilai yang hilang (NaN) menggunakan salah satu metode berikut:
- Imputasi dengan statistik deskriptif
- Imputasi dengan model

In [ ]:
# Melakukan pengecekan tabel kosong
df.isnull().sum()

Unnamed: 0                                          0
Student Id                                          0
Marital status                                     84
Application mode                                  172
Application order                                 108
Course                                            152
Daytime/evening attendance\t                       80
Previous qualification                            123
Previous qualification (grade)                     85
Nacionality                                       185
Mother's qualification                            145
Father's qualification                             96
Mother's occupation                                62
Father's occupation                               144
Admission grade                                   141
Displaced                                         122
Educational special needs                         119
Debtor                                            116
Tuition fees up to date     

Proses imputasi dasar menggunakan metode most frequence (modus) untuk mengatasi nilai NaN pada data kategorik

In [ ]:
from sklearn.impute import SimpleImputer

numeric_columns = df.select_dtypes(include=['int64', 'float64']).columns
categorical_columns = df.select_dtypes(include=['object']).columns

num_imputer = SimpleImputer(strategy = 'median')
df_num = num_imputer.fit_transform(df[numeric_columns])

cat_imputer = SimpleImputer(strategy = 'most_frequent')
df_cat = cat_imputer.fit_transform(df[categorical_columns])

Membersihkan nama kolom dari karakter spesial

In [ ]:
features.columns = (
    features.columns
    .str.replace('[^A-Za-z0-9_]+', '_', regex=True)
    .str.strip('_')                                  
)

Encoding

In [ ]:
from sklearn.preprocessing import LabelEncoder

label_encoders = {}
for col in features.select_dtypes(include=['object']).columns:
    le = LabelEncoder()
    features[col] = le.fit_transform(features[col].astype(str))
    label_encoders[col] = le

Imputasi pakai randomforest

In [ ]:
import miceforest as mf
kernel = mf.ImputationKernel(
    data = features,
    save_all_iterations_data = True,
    random_state = 42
)

kernel.mice(3)
features_imputed = kernel.complete_data()

Hasil setelah dilakukan imputasi

In [ ]:
features_imputed.isnull().sum()

Unnamed_0                                       0
Student_Id                                      0
Marital_status                                  0
Application_mode                                0
Application_order                               0
Course                                          0
Daytime_evening_attendance                      0
Previous_qualification                          0
Previous_qualification_grade                    0
Nacionality                                     0
Mother_s_qualification                          0
Father_s_qualification                          0
Mother_s_occupation                             0
Father_s_occupation                             0
Admission_grade                                 0
Displaced                                       0
Educational_special_needs                       0
Debtor                                          0
Tuition_fees_up_to_date                         0
Gender                                          0
